File: src/help_render.ts
/**
 * help_render.ts — renderers for the built-in help.
 *
 * The text renderer is the deterministic floor: no width assumptions, no
 * color, readable when redirected. The rich renderer is a safe ANSI subset:
 * colored headings, emphasized safety markers, uniform code styling (every
 * code span and block in one color), and the home splash logo. Tables and
 * examples keep the deterministic text layout — no width-aware realignment,
 * no token-level syntax highlighting, no hyperlinks, no pager.
 *
 * Every ANSI run is self-closed (`<style>…\x1b[0m`), so rich output stripped
 * of escapes equals the text output exactly.
 */
import { groupedPages, reachOf, type HelpPage } from "./help.ts";

/** `--rich` value: `auto` adapts to terminal interactivity. */
export type RichMode = "auto" | "always" | "never";

const ANSI = {
  reset: "\x1b[0m",
  /** Bold bright cyan — page headers and section headings. */
  heading: "\x1b[1;36m",
  /** Cyan — uniform code styling for spans and blocks. */
  code: "\x1b[36m",
  /** Bold yellow — emphasized safety markers in prose. */
  safety: "\x1b[1;33m",
};

/** Safety markers emphasized in prose (bold). Deterministic, conservative. */
const SAFETY_TERMS = ["rejected", "hard error", "writes nothing", "cannot", "never"];

/** The home splash (brand palette), embedded for the compiled binary.
 *  Keep in sync with brand/splash-ansi.sh. */
const SPLASH =
  "\n" +
  "  \x1b[38;2;194;65;12m❯\x1b[0m\x1b[38;2;255;255;0m❯\x1b[0m \x1b[38;2;247;244;236mlq\x1b[0m\x1b[38;2;31;111;235m ▉\x1b[0m\n" +
  "\n";

/** Render a help page for stdout, honoring the `--rich` mode. */
export function renderPage(page: HelpPage, rich: RichMode): string {
  const richOn = rich === "always" || (rich === "auto" && Deno.stdout.isTerminal());
  return richOn ? renderPageRich(page) : renderPageText(page);
}

/** Deterministic plain-text rendering of a help page — the floor. */
export function renderPageText(page: HelpPage): string {
  const out: string[] = [];
  if (page.id !== "home") {
    const header = pageHeader(page);
    out.push(header);
    out.push("=".repeat(header.length));
  }

  for (let i = 0; i < page.sections.length; i++) {
    const section = page.sections[i];
    if (!(page.id === "home" && i === 0)) out.push("");
    if (section.heading.length > 0) {
      out.push(section.heading);
      out.push("-".repeat(section.heading.length));
    }
    out.push(toPlainText(section.body));
  }

  if (page.id === "home") appendHomeMap(out, (line) => line);
  appendFurtherReading(out, page, (line) => line);

  return out.join("\n") + "\n";
}

/** ANSI rendering of the same content; ANSI-stripped it equals `renderPageText`. */
export function renderPageRich(page: HelpPage): string {
  const out: string[] = [];
  if (page.id === "home") out.push(SPLASH);

  if (page.id !== "home") {
    const header = pageHeader(page);
    out.push(`${ANSI.heading}${header}${ANSI.reset}`);
    out.push("=".repeat(header.length));
  }

  for (let i = 0; i < page.sections.length; i++) {
    const section = page.sections[i];
    if (!(page.id === "home" && i === 0)) out.push("");
    if (section.heading.length > 0) {
      out.push(`${ANSI.heading}${section.heading}${ANSI.reset}`);
      out.push("-".repeat(section.heading.length));
    }
    out.push(toStyledText(section.body));
  }

  if (page.id === "home") appendHomeMap(out, (line) => styleCodeSpan(line));
  appendFurtherReading(out, page, (line) => styleCodeSpan(line));

  return out.join("\n") + "\n";
}

/** The home page map (generated from the catalog), each line via `emit`. */
function appendHomeMap(out: string[], emit: (line: string) => string): void {
  out.push("");
  out.push(emit("Pages"));
  out.push("-".repeat("Pages".length));
  const groups = groupedPages();
  const maxReach = Math.max(
    ...groups.flatMap((g) => g.pages.map((p) => reachOf(p.id).length)),
  );
  for (const group of groups) {
    out.push(emit(`${group.group}/`));
    for (const p of group.pages) {
      out.push(emit(`  ${reachOf(p.id).padEnd(maxReach)}    lq help ${reachOf(p.id)}`));
    }
  }
}

/** The further-reading list, each line via `emit`. */
function appendFurtherReading(out: string[], page: HelpPage, emit: (line: string) => string): void {
  if (page.furtherReading.length === 0) return;
  out.push("");
  out.push(emit("Further reading"));
  out.push("-".repeat("Further reading".length));
  for (const link of page.furtherReading) {
    out.push(emit(`  lq help ${reachOf(link.page)} - ${link.hint}`));
  }
}

/** The page header line, e.g. "lq read" for commands, "lq help cst" for topics. */
function pageHeader(page: HelpPage): string {
  if (page.id === "home") return "lq";
  const reach = reachOf(page.id);
  return page.id.startsWith("commands/") ? `lq ${reach}` : `lq help ${reach}`;
}

/** Strip code markers for plain-text output: drop fence lines and inline backticks. */
function toPlainText(body: string): string {
  return body
    .split("\n")
    .filter((line) => !/^\s*```/.test(line))
    .join("\n")
    .replace(/`/g, "");
}

/** Style a body for rich output: code blocks in code color, inline code and safety markers styled. */
function toStyledText(body: string): string {
  const out: string[] = [];
  let inBlock = false;
  for (const rawLine of body.split("\n")) {
    if (/^\s*```/.test(rawLine)) {
      inBlock = !inBlock;
      continue;
    }
    if (inBlock) {
      out.push(`${ANSI.code}${rawLine}${ANSI.reset}`);
      continue;
    }
    out.push(styleInline(rawLine));
  }
  return out.join("\n");
}

/** Style one non-code line: backtick spans in code color, safety terms bold. */
function styleInline(line: string): string {
  const parts = line.split("`");
  let out = "";
  for (let i = 0; i < parts.length; i++) {
    if (i % 2 === 1) out += `${ANSI.code}${parts[i]}${ANSI.reset}`;
    else out += emphasizeSafety(parts[i]);
  }
  return out;
}

/** Bold the safety markers in prose (deterministic, self-closed runs). */
function emphasizeSafety(text: string): string {
  let out = text;
  for (const term of SAFETY_TERMS) {
    if (out.includes(term)) {
      out = out.replaceAll(term, `${ANSI.safety}${term}${ANSI.reset}`);
    }
  }
  return out;
}

/** Wrap the `lq help <page>` fragment in a line with the code color. */
function styleCodeSpan(line: string): string {
  return line.replace(/(lq help \S+)/g, `${ANSI.code}$1${ANSI.reset}`);
}

File: src/help_render.rs
//! Help page renderers (Deno `help_render.ts`).

use crate::help::{HelpPage, grouped_pages, reach_of};
use std::io::IsTerminal;

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub enum RichMode {
    Auto,
    Always,
    Never,
}

const ANSI_RESET: &str = "\x1b[0m";
const ANSI_HEADING: &str = "\x1b[1;36m";
const ANSI_CODE: &str = "\x1b[36m";
const ANSI_SAFETY: &str = "\x1b[1;33m";

const SAFETY_TERMS: &[&str] = &[
    "rejected",
    "hard error",
    "writes nothing",
    "cannot",
    "never",
];

const SPLASH: &str = concat!(
    "\n",
    "  \x1b[38;2;194;65;12m❯\x1b[0m\x1b[38;2;255;255;0m❯\x1b[0m \x1b[38;2;247;244;236mlq\x1b[0m\x1b[38;2;31;111;235m ▉\x1b[0m\n",
    "\n",
);

pub fn render_page(page: &HelpPage, rich: RichMode) -> String {
    render_page_tty(page, rich, std::io::stdout().is_terminal())
}

pub fn render_page_tty(page: &HelpPage, rich: RichMode, stdout_is_terminal: bool) -> String {
    let rich_on = match rich {
        RichMode::Always => true,
        RichMode::Never => false,
        RichMode::Auto => stdout_is_terminal,
    };
    if rich_on {
        render_page_rich(page)
    } else {
        render_page_text(page)
    }
}

pub fn render_page_text(page: &HelpPage) -> String {
    let mut out: Vec<String> = Vec::new();
    if page.id != "home" {
        out.push(page.title.to_string());
        out.push("=".repeat(page.title.len()));
    }
    for (i, section) in page.sections.iter().enumerate() {
        if !(page.id == "home" && i == 0) {
            out.push(String::new());
        }
        if !section.heading.is_empty() {
            out.push(section.heading.to_string());
            out.push("-".repeat(section.heading.len()));
        }
        out.push(to_plain_text(section.body));
    }
    if page.id == "home" {
        append_home_map(&mut out, |line| line.to_string(), |line| line.to_string());
    }
    append_further_reading(
        &mut out,
        page,
        |line| line.to_string(),
        |line| line.to_string(),
    );
    out.join("\n") + "\n"
}

pub fn render_page_rich(page: &HelpPage) -> String {
    let mut out: Vec<String> = Vec::new();
    if page.id == "home" {
        out.push(SPLASH.to_string());
    }
    if page.id != "home" {
        out.push(format!("{ANSI_HEADING}{}{ANSI_RESET}", page.title));
        out.push("=".repeat(page.title.len()));
    }
    for (i, section) in page.sections.iter().enumerate() {
        if !(page.id == "home" && i == 0) {
            out.push(String::new());
        }
        if !section.heading.is_empty() {
            out.push(format!("{ANSI_HEADING}{}{ANSI_RESET}", section.heading));
            out.push("-".repeat(section.heading.len()));
        }
        out.push(to_styled_text(section.body));
    }
    if page.id == "home" {
        append_home_map(&mut out, style_code_span, style_heading);
    }
    append_further_reading(&mut out, page, style_code_span, style_heading);
    out.join("\n") + "\n"
}

fn append_home_map(
    out: &mut Vec<String>,
    emit: fn(&str) -> String,
    emit_heading: fn(&str) -> String,
) {
    out.push(String::new());
    out.push(emit_heading("Pages"));
    out.push("-".repeat("Pages".len()));
    let groups = grouped_pages();
    let max_reach = groups
        .iter()
        .flat_map(|g| g.pages.iter().map(|p| reach_of(p.id).len()))
        .max()
        .unwrap_or(0);
    for group in groups {
        out.push(emit(&format!("{}/", group.group.as_str())));
        for p in group.pages {
            let reach = reach_of(p.id);
            out.push(emit(&format!("  {reach:<max_reach$}    lq help {reach}")));
        }
    }
}

fn append_further_reading(
    out: &mut Vec<String>,
    page: &HelpPage,
    emit: fn(&str) -> String,
    emit_heading: fn(&str) -> String,
) {
    if page.further_reading.is_empty() {
        return;
    }
    out.push(String::new());
    out.push(emit_heading("Further reading"));
    out.push("-".repeat("Further reading".len()));
    for link in page.further_reading {
        out.push(emit(&format!(
            "  lq help {} - {}",
            reach_of(link.page),
            link.hint
        )));
    }
}

fn to_plain_text(body: &str) -> String {
    body.split('\n')
        .filter(|line| !is_fence(line))
        .collect::<Vec<_>>()
        .join("\n")
        .replace('`', "")
}

fn is_fence(line: &str) -> bool {
    line.trim_start().starts_with("```")
}

fn to_styled_text(body: &str) -> String {
    let mut out = Vec::new();
    let mut in_block = false;
    for raw_line in body.split('\n') {
        if is_fence(raw_line) {
            in_block = !in_block;
            continue;
        }
        if in_block {
            out.push(format!("{ANSI_CODE}{raw_line}{ANSI_RESET}"));
        } else {
            out.push(style_inline(raw_line));
        }
    }
    out.join("\n")
}

fn style_inline(line: &str) -> String {
    let parts: Vec<&str> = line.split('`').collect();
    let mut out = String::new();
    for (i, part) in parts.iter().enumerate() {
        if i % 2 == 1 {
            out.push_str(ANSI_CODE);
            out.push_str(part);
            out.push_str(ANSI_RESET);
        } else {
            out.push_str(&emphasize_safety(part));
        }
    }
    out
}

fn emphasize_safety(text: &str) -> String {
    let mut out = text.to_string();
    for term in SAFETY_TERMS {
        if out.contains(term) {
            out = out.replace(term, &format!("{ANSI_SAFETY}{term}{ANSI_RESET}"));
        }
    }
    out
}

fn style_code_span(line: &str) -> String {
    let mut out = String::new();
    let mut rest = line;
    while let Some(idx) = rest.find("lq help ") {
        out.push_str(&rest[..idx]);
        let after = &rest[idx + "lq help ".len()..];
        let token_len = after.find(char::is_whitespace).unwrap_or(after.len());
        let frag = &rest[idx..idx + "lq help ".len() + token_len];
        out.push_str(ANSI_CODE);
        out.push_str(frag);
        out.push_str(ANSI_RESET);
        rest = &rest[idx + frag.len()..];
    }
    out.push_str(rest);
    out
}

fn style_heading(line: &str) -> String {
    format!("{ANSI_HEADING}{line}{ANSI_RESET}")
}
